### ISSUE TITLE
Why PyTorch GraphSAGE shows different performance in while running CPU vs. GPU version?

### ISSUE THREAD

**Issue by khaled-rahman** [dmlc/dgl#3289]
I was running `train_full.py` file in the directory `../examples/pytorch/graphsage` for both CPU and GPU. I saw very low accuracy for the GPU version which is outlined below.

## CPU version
I ran the following command: `python3 train_full.py --dataset cora --n-epochs 100`
Then, I got the following output which seems to be *OK*.
```
Using backend: pytorch
Namespace(aggregator_type='gcn', dataset='cora', dropout=0.5, gpu=-1, lr=0.01, n_epochs=100, n_hidden=16, n_layers=1, weight_decay=0.0005)
  NumNodes: 2708
  NumEdges: 10556
  NumFeats: 1433
  NumClasses: 7
  NumTrainingSamples: 140
  NumValidationSamples: 500
  NumTestSamples: 1000
Done loading data from cached files.
/home/morahma/.local/lib/python3.8/site-packages/dgl/data/utils.py:285: UserWarning: Property dataset.graph will be deprecated, please use dataset[0] instead.
  warnings.warn('Property {} will be deprecated, please use {} instead.'.format(old, new))
----Data statistics------'
      #Edges 10556
      #Classes 7
      #Train samples 140
      #Val samples 500
      #Test samples 1000
/home/morahma/.local/lib/python3.8/site-packages/numpy/core/fromnumeric.py:3419: RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
/home/morahma/.local/lib/python3.8/site-packages/numpy/core/_methods.py:188: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
Epoch 00000 | Time(s) nan | Loss 1.9558 | Accuracy 0.1300 | ETputs(KTEPS) nan
Epoch 00001 | Time(s) nan | Loss 1.9410 | Accuracy 0.2040 | ETputs(KTEPS) nan
Epoch 00002 | Time(s) nan | Loss 1.9328 | Accuracy 0.2220 | ETputs(KTEPS) nan
Epoch 00003 | Time(s) 0.0127 | Loss 1.9276 | Accuracy 0.3760 | ETputs(KTEPS) 831.22
Epoch 00004 | Time(s) 0.0126 | Loss 1.9123 | Accuracy 0.3260 | ETputs(KTEPS) 835.39
Epoch 00005 | Time(s) 0.0126 | Loss 1.8992 | Accuracy 0.2860 | ETputs(KTEPS) 836.76
Epoch 00006 | Time(s) 0.0126 | Loss 1.8831 | Accuracy 0.2700 | ETputs(KTEPS) 837.77
Epoch 00007 | Time(s) 0.0126 | Loss 1.8761 | Accuracy 0.2860 | ETputs(KTEPS) 835.12
Epoch 00008 | Time(s) 0.0126 | Loss 1.8592 | Accuracy 0.3280 | ETputs(KTEPS) 835.84
Epoch 00009 | Time(s) 0.0131 | Loss 1.8562 | Accuracy 0.3520 | ETputs(KTEPS) 808.08
Epoch 00010 | Time(s) 0.0136 | Loss 1.8400 | Accuracy 0.4200 | ETputs(KTEPS) 777.37
... ... ...
Epoch 00090 | Time(s) 0.0130 | Loss 0.6471 | Accuracy 0.7940 | ETputs(KTEPS) 812.61
Epoch 00091 | Time(s) 0.0130 | Loss 0.6492 | Accuracy 0.8020 | ETputs(KTEPS) 812.47
Epoch 00092 | Time(s) 0.0130 | Loss 0.6757 | Accuracy 0.8040 | ETputs(KTEPS) 812.52
Epoch 00093 | Time(s) 0.0130 | Loss 0.6337 | Accuracy 0.8060 | ETputs(KTEPS) 812.45
Epoch 00094 | Time(s) 0.0130 | Loss 0.6307 | Accuracy 0.8020 | ETputs(KTEPS) 812.76
Epoch 00095 | Time(s) 0.0130 | Loss 0.6736 | Accuracy 0.7940 | ETputs(KTEPS) 813.05
Epoch 00096 | Time(s) 0.0130 | Loss 0.6617 | Accuracy 0.7940 | ETputs(KTEPS) 813.35
Epoch 00097 | Time(s) 0.0130 | Loss 0.6649 | Accuracy 0.7980 | ETputs(KTEPS) 813.62
Epoch 00098 | Time(s) 0.0130 | Loss 0.5937 | Accuracy 0.7960 | ETputs(KTEPS) 813.38
Epoch 00099 | Time(s) 0.0130 | Loss 0.6565 | Accuracy 0.7960 | ETputs(KTEPS) 813.66

Test Accuracy 0.8110
```

## GPU version
I ran the following command: `python3 train_full.py --dataset cora --n-epochs 100 --gpu 0`
Then, I got the following output which seems *NOT* to be OK.
```
Using backend: pytorch
Namespace(aggregator_type='gcn', dataset='cora', dropout=0.5, gpu=0, lr=0.01, n_epochs=100, n_hidden=16, n_layers=1, weight_decay=0.0005)
  NumNodes: 2708
  NumEdges: 10556
  NumFeats: 1433
  NumClasses: 7
  NumTrainingSamples: 140
  NumValidationSamples: 500
  NumTestSamples: 1000
Done loading data from cached files.
/home/morahma/.local/lib/python3.8/site-packages/dgl/data/utils.py:285: UserWarning: Property dataset.graph will be deprecated, please use dataset[0] instead.
  warnings.warn('Property {} will be deprecated, please use {} instead.'.format(old, new))
----Data statistics------'
      #Edges 10556
      #Classes 7
      #Train samples 140
      #Val samples 500
      #Test samples 1000
use cuda: 0
/home/morahma/.local/lib/python3.8/site-packages/numpy/core/fromnumeric.py:3419: RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
/home/morahma/.local/lib/python3.8/site-packages/numpy/core/_methods.py:188: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
Epoch 00000 | Time(s) nan | Loss 1.9516 | Accuracy 0.2900 | ETputs(KTEPS) nan
Epoch 00001 | Time(s) nan | Loss 1.9415 | Accuracy 0.2260 | ETputs(KTEPS) nan
Epoch 00002 | Time(s) nan | Loss 1.9369 | Accuracy 0.1980 | ETputs(KTEPS) nan
Epoch 00003 | Time(s) 0.0030 | Loss 1.9315 | Accuracy 0.2140 | ETputs(KTEPS) 3518.92
Epoch 00004 | Time(s) 0.0029 | Loss 1.9249 | Accuracy 0.2380 | ETputs(KTEPS) 3672.60
Epoch 00005 | Time(s) 0.0029 | Loss 1.9210 | Accuracy 0.2580 | ETputs(KTEPS) 3640.14
Epoch 00006 | Time(s) 0.0029 | Loss 1.9100 | Accuracy 0.2840 | ETputs(KTEPS) 3582.63
Epoch 00007 | Time(s) 0.0030 | Loss 1.8999 | Accuracy 0.3260 | ETputs(KTEPS) 3550.19
Epoch 00008 | Time(s) 0.0030 | Loss 1.8963 | Accuracy 0.3600 | ETputs(KTEPS) 3565.01
Epoch 00009 | Time(s) 0.0029 | Loss 1.8767 | Accuracy 0.3680 | ETputs(KTEPS) 3589.64
Epoch 00010 | Time(s) 0.0029 | Loss 1.8543 | Accuracy 0.3380 | ETputs(KTEPS) 3606.75
... ... ...
Epoch 00090 | Time(s) 0.0032 | Loss 0.7459 | Accuracy 0.5960 | ETputs(KTEPS) 3315.84
Epoch 00091 | Time(s) 0.0032 | Loss 0.7614 | Accuracy 0.6000 | ETputs(KTEPS) 3315.64
Epoch 00092 | Time(s) 0.0032 | Loss 0.6922 | Accuracy 0.6000 | ETputs(KTEPS) 3315.38
Epoch 00093 | Time(s) 0.0032 | Loss 0.6668 | Accuracy 0.6000 | ETputs(KTEPS) 3314.96
Epoch 00094 | Time(s) 0.0032 | Loss 0.7028 | Accuracy 0.5860 | ETputs(KTEPS) 3314.59
Epoch 00095 | Time(s) 0.0032 | Loss 0.6298 | Accuracy 0.5840 | ETputs(KTEPS) 3313.35
Epoch 00096 | Time(s) 0.0032 | Loss 0.7964 | Accuracy 0.5740 | ETputs(KTEPS) 3311.98
Epoch 00097 | Time(s) 0.0032 | Loss 0.6799 | Accuracy 0.5740 | ETputs(KTEPS) 3310.68
Epoch 00098 | Time(s) 0.0032 | Loss 0.7177 | Accuracy 0.5680 | ETputs(KTEPS) 3310.74
Epoch 00099 | Time(s) 0.0032 | Loss 0.6526 | Accuracy 0.5700 | ETputs(KTEPS) 3310.58

Test Accuracy 0.5570
```

Any suggestions or thought why this is happening? It would be helpful.

**Comment by VoVAllen** [comment]
There's a sampling bug on gpu fixed recently. Could you try our nightly version by `pip install --pre dgl-cu111 -f https://data.dgl.ai/wheels-test/repo.html`?

**Comment by khaled-rahman** [comment]
Thanks so much. Yes, it's better and the accuracy of the GPU version is close to the CPU version.
